### ISSUE TITLE
error C3493: 'test_val' cannot be implicitly captured because no default capture mode has been specified

### ISSUE THREAD

**Issue by xkszltl** [pytorch/pytorch#44691]
## 🐛 Bug

Regression of the day~

Our Windows build failed today with the following compile error:
```
[4/7] "C:\PROGRA~2\Microsoft Visual Studio\2019\Enterprise\VC\Tools\MSVC\14.26.28801\bin\Hostx64\x64\cl.exe"   /TP -DBUILDING_TESTS -DFMT_HEADER_ONLY=1 -DGFLAGS_IS_A_DLL=1 -DGLOG_NO_ABBREVIATED_SEVERITIES -DGOOGLE_GLOG_DLL_DECL=__declspec(dllimport) -DGOOGLE_GLOG_DLL_DECL_FOR_UNITTESTS=__declspec(dllimport) -DIDEEP_USE_MKL -DMINIZ_DISABLE_ZIP_READER_CRC32_CHECKS -DONNXIFI_ENABLE_EXT=1 -DONNX_ML=1 -DONNX_NAMESPACE=onnx_torch -DPROTOBUF_USE_DLLS -DTHP_BUILD_MAIN_LIB -DTH_BLAS_MKL -DUSE_CUDA -DUSE_CUDNN -DUSE_EXTERNAL_MZCRC -DUSE_NUMPY -DWIN32_LEAN_AND_MEAN -D_CRT_SECURE_NO_DEPRECATE=1 -D_OPENMP_NOFORCE_MANIFEST -Dtorch_python_EXPORTS -Iaten\src -I..\aten\src -I. -I..\ -I..\cmake\..\third_party\benchmark\include -Icaffe2\contrib\aten -I..\third_party\onnx -Ithird_party\onnx -I..\third_party\foxi -Ithird_party\foxi -I..\torch\.. -I..\torch\..\aten\src -I..\torch\..\aten\src\TH -Icaffe2\aten\src -Ithird_party -I..\torch\..\third_party\gloo -I..\torch\..\third_party\onnx -I..\torch\csrc -I..\torch\csrc\api\include -I..\torch\lib -I..\torch\lib\libshm_windows -I..\torch\csrc\api -I..\c10\.. -Ithird_party\ideep\mkl-dnn\include -I..\third_party\ideep\mkl-dnn\src\..\include -I..\c10\cuda\..\.. -I..\third_party\fmt\include -I..\cmake\..\third_party\googletest\googlemock\include -I..\cmake\..\third_party\googletest\googletest\include -I"C:\Program Files\protobuf\include" -I"C:\Program Files (x86)\IntelSWTools\compilers_and_libraries\windows\mkl\include" -I..\third_party -I"C:\Program Files\opencv\include" -I..\cmake\..\third_party\eigen -I"C:\Program Files\Python38\include" -I"C:\Program Files\Python38\lib\site-packages\numpy\core\include" -I"C:\Program Files\pybind11\include" -I"C:\Program Files\NVIDIA GPU Computing Toolkit\CUDA\v11.0\include" -I..\third_party\ideep\mkl-dnn\include -I..\third_party\ideep\include -I"C:\Program Files\NVIDIA Corporation\NvToolsExt\include" -I"C:\Program Files\zlib\include" -I"C:\Program Files\gflags\include" -I"C:\Program Files\google-glog\include" /EHsc /FS /GL /MP  /arch:AVX2 /DGFLAGS_IS_A_DLL=1 /DPROTOBUF_USE_DLLS /DWIN32 /D_SILENCE_STDEXT_HASH_DEPRECATION_WARNINGS /D_SILENCE_TR1_NAMESPACE_DEPRECATION_WARNING /w /w /bigobj -openmp:experimental -DNDEBUG -DUSE_FBGEMM -DUSE_VULKAN_WRAPPER -DHAVE_AVX_CPU_DEFINITION -DHAVE_AVX2_CPU_DEFINITION /MD /O2 /Ob2 /DNDEBUG /w /bigobj -DNDEBUG -DCUDA_HAS_FP16=1 -DUSE_GCC_GET_CPUID -DUSE_AVX -DUSE_AVX2 -DTH_HAVE_THREAD /EHsc /DNOMINMAX /wd4267 /wd4251 /wd4522 /wd4838 /wd4305 /wd4244 /wd4190 /wd4101 /wd4996 /wd4275 /bigobj -std:c++14 /showIncludes /Focaffe2\torch\CMakeFiles\torch_python.dir\__\test\cpp\jit\test_misc.cpp.obj /Fdcaffe2\torch\CMakeFiles\torch_python.dir\ /FS -c ..\test\cpp\jit\test_misc.cpp
FAILED: caffe2/torch/CMakeFiles/torch_python.dir/__/test/cpp/jit/test_misc.cpp.obj
"C:\PROGRA~2\Microsoft Visual Studio\2019\Enterprise\VC\Tools\MSVC\14.26.28801\bin\Hostx64\x64\cl.exe"   /TP -DBUILDING_TESTS -DFMT_HEADER_ONLY=1 -DGFLAGS_IS_A_DLL=1 -DGLOG_NO_ABBREVIATED_SEVERITIES -DGOOGLE_GLOG_DLL_DECL=__declspec(dllimport) -DGOOGLE_GLOG_DLL_DECL_FOR_UNITTESTS=__declspec(dllimport) -DIDEEP_USE_MKL -DMINIZ_DISABLE_ZIP_READER_CRC32_CHECKS -DONNXIFI_ENABLE_EXT=1 -DONNX_ML=1 -DONNX_NAMESPACE=onnx_torch -DPROTOBUF_USE_DLLS -DTHP_BUILD_MAIN_LIB -DTH_BLAS_MKL -DUSE_CUDA -DUSE_CUDNN -DUSE_EXTERNAL_MZCRC -DUSE_NUMPY -DWIN32_LEAN_AND_MEAN -D_CRT_SECURE_NO_DEPRECATE=1 -D_OPENMP_NOFORCE_MANIFEST -Dtorch_python_EXPORTS -Iaten\src -I..\aten\src -I. -I..\ -I..\cmake\..\third_party\benchmark\include -Icaffe2\contrib\aten -I..\third_party\onnx -Ithird_party\onnx -I..\third_party\foxi -Ithird_party\foxi -I..\torch\.. -I..\torch\..\aten\src -I..\torch\..\aten\src\TH -Icaffe2\aten\src -Ithird_party -I..\torch\..\third_party\gloo -I..\torch\..\third_party\onnx -I..\torch\csrc -I..\torch\csrc\api\include -I..\torch\lib -I..\torch\lib\libshm_windows -I..\torch\csrc\api -I..\c10\.. -Ithird_party\ideep\mkl-dnn\include -I..\third_party\ideep\mkl-dnn\src\..\include -I..\c10\cuda\..\.. -I..\third_party\fmt\include -I..\cmake\..\third_party\googletest\googlemock\include -I..\cmake\..\third_party\googletest\googletest\include -I"C:\Program Files\protobuf\include" -I"C:\Program Files (x86)\IntelSWTools\compilers_and_libraries\windows\mkl\include" -I..\third_party -I"C:\Program Files\opencv\include" -I..\cmake\..\third_party\eigen -I"C:\Program Files\Python38\include" -I"C:\Program Files\Python38\lib\site-packages\numpy\core\include" -I"C:\Program Files\pybind11\include" -I"C:\Program Files\NVIDIA GPU Computing Toolkit\CUDA\v11.0\include" -I..\third_party\ideep\mkl-dnn\include -I..\third_party\ideep\include -I"C:\Program Files\NVIDIA Corporation\NvToolsExt\include" -I"C:\Program Files\zlib\include" -I"C:\Program Files\gflags\include" -I"C:\Program Files\google-glog\include" /EHsc /FS /GL /MP  /arch:AVX2 /DGFLAGS_IS_A_DLL=1 /DPROTOBUF_USE_DLLS /DWIN32 /D_SILENCE_STDEXT_HASH_DEPRECATION_WARNINGS /D_SILENCE_TR1_NAMESPACE_DEPRECATION_WARNING /w /w /bigobj -openmp:experimental -DNDEBUG -DUSE_FBGEMM -DUSE_VULKAN_WRAPPER -DHAVE_AVX_CPU_DEFINITION -DHAVE_AVX2_CPU_DEFINITION /MD /O2 /Ob2 /DNDEBUG /w /bigobj -DNDEBUG -DCUDA_HAS_FP16=1 -DUSE_GCC_GET_CPUID -DUSE_AVX -DUSE_AVX2 -DTH_HAVE_THREAD /EHsc /DNOMINMAX /wd4267 /wd4251 /wd4522 /wd4838 /wd4305 /wd4244 /wd4190 /wd4101 /wd4996 /wd4275 /bigobj -std:c++14 /showIncludes /Focaffe2\torch\CMakeFiles\torch_python.dir\__\test\cpp\jit\test_misc.cpp.obj /Fdcaffe2\torch\CMakeFiles\torch_python.dir\ /FS -c ..\test\cpp\jit\test_misc.cpp
Microsoft (R) C/C++ Optimizing Compiler Version 19.26.28806 for x64
Copyright (C) Microsoft Corporation.  All rights reserved.

..\test\cpp\jit\test_misc.cpp(1051): error C3493: 'test_val' cannot be implicitly captured because no default capture mode has been specified
..\test\cpp\jit\test_misc.cpp(1060): error C3493: 'test_val' cannot be implicitly captured because no default capture mode has been specified
..\test\cpp\jit\test_misc.cpp(1048): error C2440: '<function-style-cast>': cannot convert from 'initializer list' to 'at::RecordFunctionCallback'
..\test\cpp\jit\test_misc.cpp(1062): note: No constructor could take the source type, or constructor overload resolution was ambiguous
..\test\cpp\jit\test_misc.cpp(1048): error C2064: term does not evaluate to a function taking 2 arguments
..\test\cpp\jit\test_misc.cpp(1077): error C3493: 'test_val' cannot be implicitly captured because no default capture mode has been specified
..\test\cpp\jit\test_misc.cpp(1086): error C3493: 'test_val' cannot be implicitly captured because no default capture mode has been specified
..\test\cpp\jit\test_misc.cpp(1074): error C2440: '<function-style-cast>': cannot convert from 'initializer list' to 'at::RecordFunctionCallback'
..\test\cpp\jit\test_misc.cpp(1088): note: No constructor could take the source type, or constructor overload resolution was ambiguous
..\test\cpp\jit\test_misc.cpp(1074): error C2064: term does not evaluate to a function taking 2 arguments
[5/7] "C:\PROGRA~2\Microsoft Visual Studio\2019\Enterprise\VC\Tools\MSVC\14.26.28801\bin\Hostx64\x64\cl.exe"   /TP -DGFLAGS_IS_A_DLL=1 -DGLOG_NO_ABBREVIATED_SEVERITIES -DGOOGLE_GLOG_DLL_DECL=__declspec(dllimport) -DGOOGLE_GLOG_DLL_DECL_FOR_UNITTESTS=__declspec(dllimport) -DIDEEP_USE_MKL -DMINIZ_DISABLE_ZIP_READER_CRC32_CHECKS -DONNXIFI_ENABLE_EXT=1 -DONNX_ML=1 -DONNX_NAMESPACE=onnx_torch -DPROTOBUF_USE_DLLS -DTH_BLAS_MKL -DUSE_CUDA -DUSE_EXTERNAL_MZCRC -DWIN32_LEAN_AND_MEAN -D_CRT_SECURE_NO_DEPRECATE=1 -D_OPENMP_NOFORCE_MANIFEST -Iaten\src -I..\aten\src -I. -I..\ -I..\cmake\..\third_party\benchmark\include -Icaffe2\contrib\aten -I..\third_party\onnx -Ithird_party\onnx -I..\third_party\foxi -Ithird_party\foxi -Icaffe2\..\aten\src -Icaffe2\..\aten\src\ATen -I..\torch\csrc\api -I..\torch\csrc\api\include -I..\c10\.. -Ithird_party\ideep\mkl-dnn\include -I..\third_party\ideep\mkl-dnn\src\..\include -I..\c10\cuda\..\.. -I..\cmake\..\third_party\googletest\googlemock\include -I..\cmake\..\third_party\googletest\googletest\include -I"C:\Program Files\protobuf\include" -I"C:\Program Files (x86)\IntelSWTools\compilers_and_libraries\windows\mkl\include" -I..\third_party -I"C:\Program Files\opencv\include" -I..\cmake\..\third_party\eigen -I"C:\Program Files\Python38\include" -I"C:\Program Files\Python38\lib\site-packages\numpy\core\include" -I"C:\Program Files\pybind11\include" -I"C:\Program Files\NVIDIA GPU Computing Toolkit\CUDA\v11.0\include" -I..\third_party\ideep\mkl-dnn\include -I..\third_party\ideep\include -I"C:\Program Files\zlib\include" -I"C:\Program Files\gflags\include" -I"C:\Program Files\google-glog\include" -I"C:\Program Files\NVIDIA Corporation\NvToolsExt\include" -I..\third_party\googletest\googletest\include -I..\third_party\googletest\googletest /EHsc /FS /GL /MP  /arch:AVX2 /DGFLAGS_IS_A_DLL=1 /DPROTOBUF_USE_DLLS /DWIN32 /D_SILENCE_STDEXT_HASH_DEPRECATION_WARNINGS /D_SILENCE_TR1_NAMESPACE_DEPRECATION_WARNING /w /w /bigobj -openmp:experimental -DNDEBUG -DUSE_FBGEMM -DUSE_VULKAN_WRAPPER -DHAVE_AVX_CPU_DEFINITION -DHAVE_AVX2_CPU_DEFINITION /MD /O2 /Ob2 /DNDEBUG /w /bigobj -DNDEBUG -DCUDA_HAS_FP16=1 -DUSE_GCC_GET_CPUID -DUSE_AVX -DUSE_AVX2 -DTH_HAVE_THREAD /EHsc /DNOMINMAX /wd4267 /wd4251 /wd4522 /wd4838 /wd4305 /wd4244 /wd4190 /wd4101 /wd4996 /wd4275 /bigobj -std:c++14 /showIncludes /Fotest_jit\CMakeFiles\test_jit.dir\test_misc.cpp.obj /Fdtest_jit\CMakeFiles\test_jit.dir\ /FS -c ..\test\cpp\jit\test_misc.cpp
FAILED: test_jit/CMakeFiles/test_jit.dir/test_misc.cpp.obj
"C:\PROGRA~2\Microsoft Visual Studio\2019\Enterprise\VC\Tools\MSVC\14.26.28801\bin\Hostx64\x64\cl.exe"   /TP -DGFLAGS_IS_A_DLL=1 -DGLOG_NO_ABBREVIATED_SEVERITIES -DGOOGLE_GLOG_DLL_DECL=__declspec(dllimport) -DGOOGLE_GLOG_DLL_DECL_FOR_UNITTESTS=__declspec(dllimport) -DIDEEP_USE_MKL -DMINIZ_DISABLE_ZIP_READER_CRC32_CHECKS -DONNXIFI_ENABLE_EXT=1 -DONNX_ML=1 -DONNX_NAMESPACE=onnx_torch -DPROTOBUF_USE_DLLS -DTH_BLAS_MKL -DUSE_CUDA -DUSE_EXTERNAL_MZCRC -DWIN32_LEAN_AND_MEAN -D_CRT_SECURE_NO_DEPRECATE=1 -D_OPENMP_NOFORCE_MANIFEST -Iaten\src -I..\aten\src -I. -I..\ -I..\cmake\..\third_party\benchmark\include -Icaffe2\contrib\aten -I..\third_party\onnx -Ithird_party\onnx -I..\third_party\foxi -Ithird_party\foxi -Icaffe2\..\aten\src -Icaffe2\..\aten\src\ATen -I..\torch\csrc\api -I..\torch\csrc\api\include -I..\c10\.. -Ithird_party\ideep\mkl-dnn\include -I..\third_party\ideep\mkl-dnn\src\..\include -I..\c10\cuda\..\.. -I..\cmake\..\third_party\googletest\googlemock\include -I..\cmake\..\third_party\googletest\googletest\include -I"C:\Program Files\protobuf\include" -I"C:\Program Files (x86)\IntelSWTools\compilers_and_libraries\windows\mkl\include" -I..\third_party -I"C:\Program Files\opencv\include" -I..\cmake\..\third_party\eigen -I"C:\Program Files\Python38\include" -I"C:\Program Files\Python38\lib\site-packages\numpy\core\include" -I"C:\Program Files\pybind11\include" -I"C:\Program Files\NVIDIA GPU Computing Toolkit\CUDA\v11.0\include" -I..\third_party\ideep\mkl-dnn\include -I..\third_party\ideep\include -I"C:\Program Files\zlib\include" -I"C:\Program Files\gflags\include" -I"C:\Program Files\google-glog\include" -I"C:\Program Files\NVIDIA Corporation\NvToolsExt\include" -I..\third_party\googletest\googletest\include -I..\third_party\googletest\googletest /EHsc /FS /GL /MP  /arch:AVX2 /DGFLAGS_IS_A_DLL=1 /DPROTOBUF_USE_DLLS /DWIN32 /D_SILENCE_STDEXT_HASH_DEPRECATION_WARNINGS /D_SILENCE_TR1_NAMESPACE_DEPRECATION_WARNING /w /w /bigobj -openmp:experimental -DNDEBUG -DUSE_FBGEMM -DUSE_VULKAN_WRAPPER -DHAVE_AVX_CPU_DEFINITION -DHAVE_AVX2_CPU_DEFINITION /MD /O2 /Ob2 /DNDEBUG /w /bigobj -DNDEBUG -DCUDA_HAS_FP16=1 -DUSE_GCC_GET_CPUID -DUSE_AVX -DUSE_AVX2 -DTH_HAVE_THREAD /EHsc /DNOMINMAX /wd4267 /wd4251 /wd4522 /wd4838 /wd4305 /wd4244 /wd4190 /wd4101 /wd4996 /wd4275 /bigobj -std:c++14 /showIncludes /Fotest_jit\CMakeFiles\test_jit.dir\test_misc.cpp.obj /Fdtest_jit\CMakeFiles\test_jit.dir\ /FS -c ..\test\cpp\jit\test_misc.cpp
Microsoft (R) C/C++ Optimizing Compiler Version 19.26.28806 for x64
Copyright (C) Microsoft Corporation.  All rights reserved.

..\test\cpp\jit\test_misc.cpp(1051): error C3493: 'test_val' cannot be implicitly captured because no default capture mode has been specified
..\test\cpp\jit\test_misc.cpp(1060): error C3493: 'test_val' cannot be implicitly captured because no default capture mode has been specified
..\test\cpp\jit\test_misc.cpp(1048): error C2440: '<function-style-cast>': cannot convert from 'initializer list' to 'at::RecordFunctionCallback'
..\test\cpp\jit\test_misc.cpp(1062): note: No constructor could take the source type, or constructor overload resolution was ambiguous
..\test\cpp\jit\test_misc.cpp(1048): error C2064: term does not evaluate to a function taking 2 arguments
..\test\cpp\jit\test_misc.cpp(1077): error C3493: 'test_val' cannot be implicitly captured because no default capture mode has been specified
..\test\cpp\jit\test_misc.cpp(1086): error C3493: 'test_val' cannot be implicitly captured because no default capture mode has been specified
..\test\cpp\jit\test_misc.cpp(1074): error C2440: '<function-style-cast>': cannot convert from 'initializer list' to 'at::RecordFunctionCallback'
..\test\cpp\jit\test_misc.cpp(1088): note: No constructor could take the source type, or constructor overload resolution was ambiguous
..\test\cpp\jit\test_misc.cpp(1074): error C2064: term does not evaluate to a function taking 2 arguments
ninja: build stopped: subcommand failed.
```

## Environment

 - PyTorch Version (e.g., 1.0): master
 - OS (e.g., Linux): Win Server 2019
 - How you installed PyTorch (`conda`, `pip`, source): source
 - Build command you used (if compiling from source): cmake+ninja+vs2019
 - Python version: 3.8.5
 - CUDA/cuDNN version: 11.0/8

**Comment by xkszltl** [comment]
@louisfeng 
FYI this is from your commit: https://github.com/pytorch/pytorch/commit/71673b31f932b671ace0b4d44a559891f770396e#diff-d60554f530a96690b217bc591f0fb52e

**Comment by ngimel** [comment]
This PR has been reverted

**Comment by xkszltl** [comment]
@ngimel 
I only found the PR above it (the green one) is reverted, not the one I mentioned (red):
<img width="1272" alt="image" src="https://user-images.githubusercontent.com/5203025/93174816-3e758900-f761-11ea-9749-6fc58b556704.png">
Latest commit to the broken file remains unchanged:
![Uploading image.png…]()

**Comment by xkszltl** [comment]
Anyway, I've sent a PR to fix it.

**Comment by louisfeng** [comment]
@xkszltl thanks for looking into this issue. This is bug in VS2019 where it's not properly capturing constant integral value. Where as other compilers behave properly. This bug is supposedly fixed in newer version of VS compiler 16.3, see https://developercommunity.visualstudio.com/content/problem/367326/problems-with-capturing-constexpr-in-lambda.html

**Comment by louisfeng** [comment]
> This PR has been reverted

@ngimel That's not the right PR and should not be reverted.
